src/type_icons.c

```diff
@@ -473,13 +473,24 @@ static void FreeAllTypeIconResources(void)
 
 static bool32 ShouldHideTypeIcon(u32 battlerId)
 {
-    return gBattlerControllerFuncs[battlerId] != PlayerHandleChooseMove
-        && gBattlerControllerFuncs[battlerId] != HandleInputChooseMove
-        && gBattlerControllerFuncs[battlerId] != HandleChooseMoveAfterDma3
-        && gBattlerControllerFuncs[battlerId] != HandleInputChooseMove
-        && gBattlerControllerFuncs[battlerId] != HandleInputChooseTarget
-        && gBattlerControllerFuncs[battlerId] != HandleMoveSwitching
-        && gBattlerControllerFuncs[battlerId] != HandleInputChooseMove;
+    u32 funcIndex;
+
+    void* showTypesControllerFuncs[] =
+    {
+        PlayerHandleChooseMove,
+        HandleChooseMoveAfterDma3,
+        HandleInputChooseTarget,
+        HandleInputShowTargets,
+        HandleInputShowEntireFieldTargets,
+        HandleMoveSwitching,
+        HandleInputChooseMove,
+    };
```

Comment:
If I'm not mistaken, since this array won't change, it should be a const array outside of the function, to optimize the function.

Reply:
I wonder if the lag would be visible if it has to loop 7 times. This would be 7^2, right?

Reply:
oh array count is a pre proc so it doesn't matter

Reply:
According to @hedara90 `ARRAY_COUNT` is a preproc so a define is technically not needed here.

Reply:
@AlexOn1ine 

1) It has to iterate through the full list to return `FALSE`, so it already does. I legitimately don't know if a `for` loop is more computationally expensive than the 7 `if` statements that it replaced.

2) I'm confused by [this](https://github.com/rh-hideout/pokeemerald-expansion/pull/5201#discussion_r1730352555), are you saying to discard Edu's `const array` feedback?

Reply:
> It has to iterate through the full list to return FALSE, so it already does. I legitimately don't know if a for loop is more computationally expensive than the 7 if statements that it replaced.

That wasn't the problem. We didn't know if `ARRAY_COUNT` would be executed each time a loop is done. which would mean it will be 7^2. I think that is why Edu brought up the constant but afaik we use `ARRAY_COUNT` everywhere generally. It also seems to be a preproc according to hedara.

Reply:
```C
#define ARRAY_COUNT(array) (size_t)(sizeof(array) / sizeof((array)[0]))

int switchThing(int input)
{
    int myArr[4] = {1, 2, 3, 4};
    int retVal = input + ARRAY_COUNT(myArr);
    return retVal;
}
```
Compiles down to
```asm
switchThing:
        adds    r0, r0, #4
        bx      lr
```

Reply:
I think I understand now.

So that's a "don't touch it, we're good"?

Reply:
would have to ask @AsparagusEduardo

Reply:
I was not talking about ARRAY_COUNT 😅
What I meant is if `showTypesControllerFuncs` is always going to be the same, doesn't it make sense for it to be defined outside of the function, instead of it being it being created as part of the function's flow?

Reply:
my bad

Reply:
I'm failing a skill check.

```c
static const void* showTypesControllerFuncs[] =
{
    PlayerHandleChooseMove,
    HandleChooseMoveAfterDma3,
    HandleInputChooseTarget,
    HandleInputShowTargets,
    HandleInputShowEntireFieldTargets,
    HandleMoveSwitching,
    HandleInputChooseMove,
};

static bool32 ShouldHideTypeIcon(u32 battlerId)
{
    u32 funcIndex;

    for (funcIndex = 0; funcIndex < ARRAY_COUNT(showTypesControllerFuncs); funcIndex++)
        if (gBattlerControllerFuncs[battlerId] == showTypesControllerFuncs[funcIndex])
            return FALSE;

    return TRUE;
}
```
yields

```
arm-none-eabi-ld: warning: ../../pokeemerald.elf has a LOAD segment with RWX permiss
ions
`.data' referenced in section `.text' of src/type_icons.o: defined in discarded sect
ion `.data' of src/type_icons.o
`.data' referenced in section `.text' of src/type_icons.o: defined in discarded sect
ion `.data' of src/type_icons.o
make: *** [Makefile:430: pokeemerald.elf] Error 1
make: Target 'all' not remade because of errors.
```

Where else in the repo is there an array of void pointers? I'll try to copy from there.

Reply:
https://github.com/rh-hideout/pokeemerald-expansion/pull/5201